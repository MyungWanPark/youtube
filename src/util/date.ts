import { format, register } from 'timeago.js';
import koLocale from 'timeago.js/lib/lang/ko.js';

register('ko', koLocale);

export function formatAgo(date: string, lang: string = 'en_US') {
    return format(date, lang);
}
