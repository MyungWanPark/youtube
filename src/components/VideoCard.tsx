import React from 'react';
import { VideoItem } from '../types/videoType';
import { formatAgo } from './../util/date';
import { useNavigate } from 'react-router-dom';

type Props = {
    video: VideoItem;
    type?: 'list';
};

export default function VideoCard({ video, type }: Props) {
    const { publishedAt, title, thumbnails, channelTitle } = video.snippet;
    const navigate = useNavigate();
    const isList = type === 'list';

    return (
        <li
            className={`${isList ? 'flex m-2' : ''} cursor-pointer`}
            onClick={() => navigate(`/videos/watch/${video.id}`, { state: { video } })}
        >
            <img className={isList ? 'w-60 mr-3' : 'w-full'} src={thumbnails.medium.url} alt={title} />
            <div>
                <p className="font-semibold my-2 line-clamp-2">{title}</p>
                <p className="text-sm opacity-80">{channelTitle}</p>
                <p className="text-sm opacity-80">{formatAgo(publishedAt)}</p>
            </div>
        </li>
    );
}
