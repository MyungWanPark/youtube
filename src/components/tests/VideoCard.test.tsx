import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import VideoCard from '../VideoCard';
import { formatAgo } from './../../util/date';

describe('VideoCard', () => {
    const video = {
        id: '8zIf0XvoL9Y',
        snippet: {
            publishedAt: '2022-12-16T14:00:08Z',
            channelId: 'UCjmJDM5pRKbUlVIzDYYWb6g',
            title: 'Barbie | Teaser Trailer',
            description: '#BarbieTheMovie, from director Greta Gerwig, only in theatres Summer 2023',
            thumbnails: {
                default: {
                    url: 'https://i.ytimg.com/vi/8zIf0XvoL9Y/default.jpg',
                    width: 120,
                    height: 90,
                },
                medium: {
                    url: 'https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg',
                    width: 320,
                    height: 180,
                },
            },
            channelTitle: 'Warner Bros. Pictures',
        },
    };
    const { publishedAt, title, thumbnails, channelTitle } = video.snippet;
    test('render video items', () => {
        //given
        render(
            <MemoryRouter>
                <VideoCard video={video} />
            </MemoryRouter>
        );

        const img = screen.getByRole('img') as HTMLImageElement;
        expect(img.src).toBe(thumbnails.medium.url);
        expect(img.alt).toBe(title);
        expect(screen.getByText(title)).toBeInTheDocument();
        expect(screen.getByText(channelTitle)).toBeInTheDocument();
        expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
    });
});
