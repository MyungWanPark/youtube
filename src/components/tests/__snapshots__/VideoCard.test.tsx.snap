// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard render video items with grid type 1`] = `
<li
  className=" cursor-pointer"
  onClick={[Function]}
>
  <img
    alt="Barbie | Teaser Trailer"
    className="w-full"
    src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      Barbie | Teaser Trailer
    </p>
    <p
      className="text-sm opacity-80"
    >
      Warner Bros. Pictures
    </p>
    <p
      className="text-sm opacity-80"
    >
      3 months ago
    </p>
  </div>
</li>
`;

exports[`VideoCard render video items with list type 1`] = `
<li
  className="flex m-2 cursor-pointer"
  onClick={[Function]}
>
  <img
    alt="Barbie | Teaser Trailer"
    className="w-60 mr-3"
    src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      Barbie | Teaser Trailer
    </p>
    <p
      className="text-sm opacity-80"
    >
      Warner Bros. Pictures
    </p>
    <p
      className="text-sm opacity-80"
    >
      3 months ago
    </p>
  </div>
</li>
`;
